File: android/app/src/main/java/com/sharesync/android/sync/SyncModels.kt
package com.sharesync.android.sync

data class PairingPayload(
    val version: Int = 1,
    val type: String = "sharesync_pairing",
    val deviceId: String,
    val deviceName: String,
    val platform: String = "android",
    val publicKey: String,
    val ip: String,
    val port: Int,
    val pairingToken: String,
    val expiresAt: String,
)

data class SyncManifest(
    val version: Int = 1,
    val sourceDeviceId: String,
    val generatedAt: String,
    val cursor: String,
    val media: List<MediaAsset>,
    val contacts: List<ContactItem> = emptyList(),
    val files: List<FileItem> = emptyList(),
)

data class MediaAsset(
    val assetId: String,
    val sourceDeviceId: String,
    val mediaType: MediaType,
    val fileName: String,
    val mimeType: String,
    val size: Long,
    val sha256: String? = null,
    val createdAt: String? = null,
    val modifiedAt: String? = null,
    val takenAt: String? = null,
    val width: Int? = null,
    val height: Int? = null,
    val durationMs: Long? = null,
    val relativePath: String? = null,
)

enum class MediaType {
    photo,
    video,
}

data class ContactItem(
    val contactId: String,
    val sourceDeviceId: String,
    val displayName: String,
    val phones: List<String>,
    val emails: List<String>,
)

data class FileItem(
    val fileId: String,
    val sourceDeviceId: String,
    val fileName: String,
    val relativePath: String,
    val size: Long,
    val mimeType: String? = null,
    val sha256: String? = null,
)

data class SyncResult(
    val syncBatchId: String,
    val targetDeviceId: String,
    val results: List<SyncItemResult>,
)

data class SyncItemResult(
    val itemType: SyncItemType,
    val sourceItemId: String,
    val targetItemId: String?,
    val status: SyncItemStatus,
    val errorCode: String? = null,
)

enum class SyncItemType {
    media,
    contact,
    file,
}

enum class SyncItemStatus {
    synced,
    skipped,
    failed,
    conflicted,
}

File: ios/ShareSync/Sync/SyncModels.swift
import Foundation

struct PairingPayload: Codable, Equatable {
    let version: Int
    let type: String
    let deviceId: String
    let deviceName: String
    let platform: String
    let publicKey: String
    let ip: String
    let port: Int
    let pairingToken: String
    let expiresAt: Date
}

struct SyncManifest: Codable, Equatable {
    let version: Int
    let sourceDeviceId: String
    let generatedAt: Date
    let cursor: String
    let media: [MediaAsset]
    let contacts: [ContactItem]
    let files: [FileItem]
}

struct MediaAsset: Codable, Equatable, Identifiable {
    var id: String { assetId }

    let assetId: String
    let sourceDeviceId: String
    let mediaType: MediaType
    let fileName: String
    let mimeType: String
    let size: Int64
    let sha256: String?
    let createdAt: Date?
    let modifiedAt: Date?
    let takenAt: Date?
    let width: Int?
    let height: Int?
    let durationMs: Int64?
    let relativePath: String?
}

enum MediaType: String, Codable {
    case photo
    case video
}

struct ContactItem: Codable, Equatable, Identifiable {
    var id: String { contactId }

    let contactId: String
    let sourceDeviceId: String
    let displayName: String
    let phones: [String]
    let emails: [String]
}

struct FileItem: Codable, Equatable, Identifiable {
    var id: String { fileId }

    let fileId: String
    let sourceDeviceId: String
    let fileName: String
    let relativePath: String
    let size: Int64
    let mimeType: String?
    let sha256: String?
}

struct SyncResult: Codable, Equatable {
    let syncBatchId: String
    let targetDeviceId: String
    let results: [SyncItemResult]
}

struct SyncItemResult: Codable, Equatable {
    let itemType: SyncItemType
    let sourceItemId: String
    let targetItemId: String?
    let status: SyncItemStatus
    let errorCode: String?
}

enum SyncItemType: String, Codable {
    case media
    case contact
    case file
}

enum SyncItemStatus: String, Codable {
    case synced
    case skipped
    case failed
    case conflicted
}

struct PhotoSyncReadiness: Equatable {
    enum BlockingReason: Equatable {
        case pairingRequired
        case endpointMissing
        case invalidPort
        case transferActive
        case photosPermissionBlocked(PhotoLibraryPermissionStatus)
    }

    enum PrimaryAction: Equatable {
        case pairAndroid
        case enterEndpoint
        case allowPhotos
        case waitForTransfer
        case fetchManifest
        case syncAllPhotos
    }

    let canFetchManifest: Bool
    let canSyncAllPhotos: Bool
    let primaryAction: PrimaryAction
    let blockingReason: BlockingReason?

    static func evaluate(
        hasPairedDevice: Bool,
        host: String,
        port: String,
        photoLibraryPermissionStatus: PhotoLibraryPermissionStatus,
        isFetchingManifest: Bool,
        isTransferActive: Bool,
        hasManifest: Bool
    ) -> PhotoSyncReadiness {
        let trimmedHost = host.trimmingCharacters(in: .whitespacesAndNewlines)
        let parsedPort = Int(port)
        let hasValidPort = parsedPort.map { (1...65_535).contains($0) } == true
        let hasEndpoint = !trimmedHost.isEmpty

        if isTransferActive {
            return PhotoSyncReadiness(
                canFetchManifest: false,
                canSyncAllPhotos: false,
                primaryAction: .waitForTransfer,
                blockingReason: .transferActive
            )
        }

        if !hasPairedDevice && !hasEndpoint {
            return PhotoSyncReadiness(
                canFetchManifest: false,
                canSyncAllPhotos: false,
                primaryAction: .pairAndroid,
                blockingReason: .pairingRequired
            )
        }

        if !hasEndpoint {
            return PhotoSyncReadiness(
                canFetchManifest: false,
                canSyncAllPhotos: false,
                primaryAction: .enterEndpoint,
                blockingReason: .endpointMissing
            )
        }

        if !hasValidPort {
            return PhotoSyncReadiness(
                canFetchManifest: false,
                canSyncAllPhotos: false,
                primaryAction: .enterEndpoint,
                blockingReason: .invalidPort
            )
        }

        if photoLibraryPermissionStatus == .denied || photoLibraryPermissionStatus == .restricted {
            return PhotoSyncReadiness(
                canFetchManifest: !isFetchingManifest,
                canSyncAllPhotos: false,
                primaryAction: .allowPhotos,
                blockingReason: .photosPermissionBlocked(photoLibraryPermissionStatus)
            )
        }

        return PhotoSyncReadiness(
            canFetchManifest: !isFetchingManifest,
            canSyncAllPhotos: !isFetchingManifest,
            primaryAction: hasManifest ? .syncAllPhotos : .fetchManifest,
            blockingReason: nil
        )
    }
}
